SPEC-337: Deprecate modifiers and clean up existing wording for defau…

craiggwilson opened the pull request:
…lt values.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mongodb/specifications/102)

<!-- Reviewable:end -->

jmikola commented on source/crud/crud.rst:
```diff
@@ -199,17 +199,25 @@ Read
+     * Optionally specifies a collation to use in MongoDB 3.4 and higher.
+     *
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.4, the driver MUST raise an error if the caller explicitly provides a value.
```
Why is raising an error required in this case, but not for `bypassDocumentValidation`? Might that be because this case entails some loss of information from the user, and `bypassDocumentValidation` is a passive option where ignoring it for older servers would be functionally equivalent behavior?

craiggwilson replied:
Yes. Not using a collation when the user specified one could result in deleting the wrong documents, for example.

jmikola commented on source/crud/crud.rst:
```diff
@@ -336,165 +338,209 @@ Read
-     *
-     * In servers < 3.2, the Partial OP_QUERY flag defaults to false.
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.2, the Partial wire protocol flag defaults to false.
```
No other text for this option refers to "Partial wire protocol flag" and this is the only indication that it relates to the wire protocol. Would it be better to omit this line or say that the effective behavior on an older server is false without referring to the wire protocol?

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -336,165 +338,209 @@ Read
+     * For servers < 3.4, the driver MUST raise an error if the caller explicitly provides a value.
      *
-     * @see http://docs.mongodb.org/manual/reference/operator/meta/comment/
+     * @see http://docs.mongodb.org/manual/reference/command/distinct/
```
Should this be https?

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -336,165 +338,209 @@ Read
-     * The default in servers >= 3.2 is no value: no "oplogReplay" option is sent with
-     * the "find" command.
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.2, the OplogReplay wire protocol flag defaults to false.
```
Same as my comment for `allowPartialResults`. Should we explicitly say that this sets the OplogReplay wire protocol flag?

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -336,165 +338,209 @@ Read
 .. code:: typescript
 
-  function calculateFirstNumberToReturn(FindOptions options) {
+  function calculateFirstNumberToReturn(FindOptions options) Int32 {
```
Should there be a colon before the return type?

craiggwilson replied:
depends on what language we are using :). I changed this and the argument as well, to typescript-ish

jmikola commented on source/crud/crud.rst:
```diff
@@ -618,38 +663,30 @@ Basic
-     *
-     * On servers < 3.2, this option is ignored.
+     * For servers >= 3.2, this option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.2, this option is ignored and not sent as documentation validation is not available.
```
"documentation" should be "document"

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -1129,106 +1173,137 @@ Find And Modify
      *
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     *
+     * Note: this option is mapped to the "new" findAndModify boolean field. ReturnDocument.Before is represents false, 
```
"is represents" is a typo.

craiggwilson replied:
done.

craiggwilson commented:
Ok, cleaned up Jeremy's comments. Also fixed up documentation for bypassDocumentValidation related to unacknowledged writes and corrected collation inaccuracies.

jmikola commented on source/crud/crud.rst:
```diff
@@ -54,10 +54,10 @@ Iterable
 --------
 
-Documentation
+documention
```
I think you search-and-replaced too much when correcting the original documentation/document typo. Some of these instances should be "documentation".

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -54,10 +54,10 @@ Iterable
 -------------
 
-The documentation provided in code below is merely for driver authors and SHOULD NOT be taken as required documentation for the driver.
+The documention provided in code below is merely for driver authors and SHOULD NOT be taken as required documention for the driver.
```
"documentation"

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -89,7 +89,7 @@ A non-exhaustive list of acceptable deviations are as follows:
   When using a fluent-style builder, all options should be named rather than inventing a new word to include in the pipeline (like options). Required parameters are still required to be on the initiating method.
 
-  In addition, it is imperative that documentation indicate when the order of operations is important. For instance, skip and limit in find is order irrelevant where skip and limit in aggregate is not.
+  In addition, it is imperative that documention indicate when the order of operations is important. For instance, skip and limit in find is order irrelevant where skip and limit in aggregate is not.
```
"documentation"

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -766,45 +777,73 @@ Bulk Write Models
+     * Specifies a collation.
+     *
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.4, the driver MUST raise an error if the caller explicitly provides a value.
```
Does this also need a "For unacknowledged writes using opcodes..." line?

craiggwilson replied:
I didn't think so, but probably. These ones are kinda weird, but better to have it there than not.

jmikola commented on source/crud/crud.rst:
```diff
@@ -766,45 +777,73 @@ Bulk Write Models
+     * Specifies a collation.
+     *
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.4, the driver MUST raise an error if the caller explicitly provides a value.
```
Does this also need a "For unacknowledged writes using opcodes..." line?

jmikola commented on source/crud/crud.rst:
```diff
@@ -766,45 +777,73 @@ Bulk Write Models
+     * Specifies a collation.
+     *
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.4, the driver MUST raise an error if the caller explicitly provides a value.
```
Does this also need a "For unacknowledged writes using opcodes..." line?

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -815,23 +854,33 @@ Bulk Write Models
+     * Specifies a collation.
+     *
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.4, the driver MUST raise an error if the caller explicitly provides a value.
```
Does this also need a "For unacknowledged writes using opcodes..." line?

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -841,23 +890,33 @@ Bulk Write Models
+     * Specifies a collation.
+     *
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.4, the driver MUST raise an error if the caller explicitly provides a value.
```
Does this also need a "For unacknowledged writes using opcodes..." line?

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -1378,13 +1437,13 @@ In addition, we have constructed some example usages in different languages that
 ==========
 
-Current drivers have chosen slightly different names and semantics for the same operations and options. In addition, not all drivers offer all the same operations and methods. As such, it is difficult to transition from driver to driver making the jobs of polyglot developers, documentation authors, and support engineers more difficult.
+Current drivers have chosen slightly different names and semantics for the same operations and options. In addition, not all drivers offer all the same operations and methods. As such, it is difficult to transition from driver to driver making the jobs of polyglot developers, documention authors, and support engineers more difficult.
```
"documentation"

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -1378,13 +1437,13 @@ In addition, we have constructed some example usages in different languages that
 =======================
 
-This spec should be mostly backwards compatible as it is very lenient. Drivers finding a backwards compatibility problem should attempt to work around it using an acceptable deviation. In rare cases, a driver may need to break backwards compatibility. This should be done in accordance with a versioning scheme indicating that a backwards compatible break may have occured in conjunction with releas ...
+This spec should be mostly backwards compatible as it is very lenient. Drivers finding a backwards compatibility problem should attempt to work around it using an acceptable deviation. In rare cases, a driver may need to break backwards compatibility. This should be done in accordance with a versioning scheme indicating that a backwards compatible break may have occured in conjunction with releas ...
```
"documentation"

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -1407,9 +1466,9 @@ Q & A
 
 Q: Why do the names of the fields differ from those defined on docs.mongodb.org?
-  Documentation and commands often refer to same-purposed fields with different names making it difficult to have a cohesive API. In addition, occasionally the name was correct at one point and its purpose has expanded to a point where the initial name doesn't accurately describe its current function.
+  documention and commands often refer to same-purposed fields with different names making it difficult to have a cohesive API. In addition, occasionally the name was correct at one point and its purpose has expanded to a point where the initial name doesn't accurately describe its current function.
```
"documentation"

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -1407,9 +1466,9 @@ Q & A
+  documention and commands often refer to same-purposed fields with different names making it difficult to have a cohesive API. In addition, occasionally the name was correct at one point and its purpose has expanded to a point where the initial name doesn't accurately describe its current function.
 
-  In addition, responses from the servers are sometimes cryptic and used for the purposes of compactness. In these cases, we felt the more verbose form was desirable for self-documentation purposes.
+  In addition, responses from the servers are sometimes cryptic and used for the purposes of compactness. In these cases, we felt the more verbose form was desirable for self-documention purposes.
```
"documentation"

craiggwilson replied:
done.

jmikola commented on source/crud/crud.rst:
```diff
@@ -1455,7 +1514,9 @@ Changes
 =======
 
 * 2016-08-17: Removed modifiers from FindOptions and added in all options.
+* 2016-08-17: Changed the value type of FindOptions.skip and FindOptions.limit to Int64 with a note related calculating batchSize for opcode writes.
```
"a note related calculating" seems to be missing a word.

craiggwilson replied:
done.

craiggwilson commented:
Comments addressed.

jyemin commented:
Nit: Since you're updating the docs URLs anyway, change the host names to docs.mongodb.com, as that's where they all redirect to anyway.

jmikola commented on source/crud/crud.rst:
```diff
@@ -123,7 +123,7 @@ Read
 .. note::
     
-    The term Iterable<T> is used below to indicate many of T. This spec is flexible on what that means as different drivers will have different requirements, types, and idioms.
+    The term Iterable<T> is used below to indicate many of T. This spec is flexible on what that means as different drivers will have different requirements, types, and idioms. 
```
Remove trailing whitespace. Should this also use an 80-column line wrap?

jmikola commented on source/crud/crud.rst:
```diff
@@ -137,21 +137,21 @@ Read
      * that if a user were to iterate a pipeline including $out, results would be returned.
      *
-     * @see http://docs.mongodb.org/manual/reference/command/aggregate/
+     * @see https://docs.mongodb.org/manual/reference/command/aggregate/
```
As @jyemin mentioned, these URLs should all use `https://docs.mongodb.com/`.

jmikola commented on source/crud/crud.rst:
```diff
@@ -336,165 +340,209 @@ Read
-     * The default in servers >= 3.2 is no value: no "awaitData" or "tailable"
-     * option is sent with the "find" command.
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     * For servers < 3.2, the AwaitData and Tailable wire protocol flag are used and default to false.
```
Pluralize "flags" in "AwaitData and Tailable wire protocol flags".

jmikola commented on source/crud/crud.rst:
```diff
@@ -336,165 +340,209 @@ Read
 
-    /**
-     * Optionally specifies a collation to use in MongoDB 3.4 and higher.
+    /** 
```
Trailing whitespace. This isn't the first instance, so you may want to go through the file and trim any occurrences.

jmikola commented on source/crud/crud.rst:
```diff
@@ -1129,148 +1237,183 @@ Find And Modify
-     * @see http://docs.mongodb.org/manual/tutorial/project-fields-from-query-results
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
+     *
+     * Note: this option is mapped to the "fields" findAndModify field.
```
Would it make more sense to say "findAndModify command option" instead of "findAndModify field"?

jmikola commented on source/crud/crud.rst:
```diff
@@ -1129,148 +1237,183 @@ Find And Modify
+     * This option is sent only if the caller explicitly provides a value. The default is to not send a value.
      *
-     * @see http://docs.mongodb.org/manual/reference/command/findAndModify/
+     * Note: this option is mapped to the "new" findAndModify boolean field. ReturnDocument.Before is represents false, 
```
Another place where we might want to say "command option" instead of "field".

Also, the "is" is a typo in "ReturnDocument.Before is represents false".

jmikola commented on source/crud/crud.rst:
```diff
@@ -1328,11 +1471,10 @@ Q & A
 
 Q: Why do the names of the fields differ from those defined on docs.mongodb.org?
-  Documentation and commands often refer to same-purposed fields with different names making it difficult to have a cohesive API. In addition, occasionally the name was correct at one point and its purpose has expanded to a point where the initial name doesn't accurately describe its current function.
+  documentation and commands often refer to same-purposed fields with different names making it difficult to have a cohesive API. In addition, occasionally the name was correct at one point and its purpose has expanded to a point where the initial name doesn't accurately describe its current function.
```
This should still be capitalized, as it's the start of a sentence.

Also, the line above refers to "docs.mongodb.org". You may want to change that to "the MongoDB manual".

jmikola commented:
See https://github.com/mongodb/specifications/pull/127.
